src/relay/pass/type_infer.cc

```diff
@@ -390,17 +432,29 @@ class TypeInferencer : private ExprFunctor<Type(const Expr&)> {
     auto ret = FuncTypeNode::make(arg_types, rtype, f->type_params, {});
     return solver_.Resolve(ret);
   }
+
+  Type VisitExpr_(const ConstructorNode* c) final {
```

Comment:
@MarisaKirisame, are you sure this is the inference rule we want? I would assume that if we're giving constructors a func type it would be like this, since the ADT's type vars would have to be type vars for the function:
```FuncType(con->inp, TypeCall(con->belong_to, con->inp), con->belong_to->tv, {})```

Reply:
Take Nil as an example. con->inp will be an empty list, while con->belong_to->tv will be a type variable.
TypeCall(List, {}) will be a kind error.